### PR TITLE
Feature target metadata

### PR DESCRIPTION
**Pull request by bogdan-bondar**
* Model for target matadata and the corresponding repository layer/management
* Target metadata quotas
* Target Mgmt API to allow CRUD operations for metadata
* Migration scripts for supported databases
* Tests for target management, Mgmt API, target metadata rsql and REST documentation

### CODE REVIEW

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 1]
```diff
@@ -146,6 +167,121 @@ public long count() {
...
+                TargetMetadata.class, Target.class, targetMetadataRepository::countByTargetId);
+    }
+
+    private JpaTarget touch(final JpaTarget target) {
+
+        // merge base target so optLockRevision gets updated and audit
+        // log written because modifying metadata is modifying the base
+        // target itself for auditing purposes.
+        final JpaTarget result = entityManager.merge(target);
+        result.setLastModifiedAt(0L);
```
[the remark this answers is not in the document]
Comment (stefbehl) [reply hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: This is done for distribution sets and software modules too.

**Comment by schabdo** [comment]
@hawkbit-bot verify please

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/EntityNotFoundException.java [hunk 2]
```diff
@@ -91,6 +91,20 @@ public EntityNotFoundException(final Class<? extends MetaData> type, final Long
...
+     * Parameterized constructor for {@link MetaData} not found.
+     * 
+     * @param type
+     *            of the entity that was not found
+     * @param enityId
+     *            of the {@link BaseEntity} the {@link MetaData} was for
+     * @param key
+     *            for the {@link MetaData} entry
+     */
+    public EntityNotFoundException(final Class<? extends MetaData> type, final String enityId, final String key) {
```
[the remark this answers is not in the document]
Comment (stefbehl) [reply hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/EntityNotFoundException.java]: fixed

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/EntityNotFoundException.java [hunk 3]
```diff
@@ -70,11 +70,25 @@ protected EntityNotFoundException(final String message) {
...
      *            of the entity that was not found
      * 
-     * @param enityId
+     * @param entityId
      *            of the {@link BaseEntity}
      */
-    public EntityNotFoundException(final Class<? extends BaseEntity> type, final Object enityId) {
-        this(type.getSimpleName() + " with given identifier {" + enityId + "} does not exist.");
+    public EntityNotFoundException(final Class<? extends BaseEntity> type, final Object entityId) {
+        this(type.getSimpleName() + " with given identifier {" + entityId + "} does not exist.");
```
[the remark this answers is not in the document]
Comment (stefbehl) [reply hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/EntityNotFoundException.java]: Ignored. Does not make sense to introduce a constant for this special literal.

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/TargetMetadataCompositeKey.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        this.target = target;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(target, key);
+    }
+
+    @Override
+    public boolean equals(final Object obj) { // NOSONAR - as this is generated
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/model/TargetMetadataCompositeKey.java]: Usually we don't use `NOSONAR` anymore. Have a look at `DistributionSetTypeElementCompositeKey` I think you don't need it anymore

**Pull request merged**